---
title: "Analysis of mutations in GZL exomes"
author: "Stacy Hung"
output:
  pdf_document: default
  html_notebook: default
---

To avoid memory errors, files have been pre-filtered using the following thresholds:
* Mutect  - predictions that have a "KEEP" flag (i.e. remove those that have a "REJECT" flag)
* Strelka - "passed" predictions (passed.somatic.snvs.vcf, passed.somatic.indels.vcf)
* VarScan - removal of predictions labelled with the effect "INTRON", "INTRAGENIC", "\tSYNONYMOUS_CODING"

Additional filters that should be applied to VarScan calls (through R):
1. GMAF (if available) < 0.01
2. predicted effect on protein (e.g. STOP_GAINED, NON_SYNONYMOUS_CODING, etc.)

OPTIMIZED thresholds (*based on ROC analysis) + additional filters:
*1. Variant allele frequency in the normal < 15%
*2. Somatic p-value < 0.175 (should be filtered already)
3. Variant reads in tumor >= 3 (should be filtered already)
4. Variant allele frequency in tumor >= 5% (should be filtered already)

Other changes to ensure consistency of tumor IDs (auto-)generated by each tool:
- Since MuTect tumor ids were truncated to not include the ".markDups" suffix, this needed to be removed from the VarScan and Strelka tumor ids

NB: Unlike in previous variant calling pipelines, the "somatic status" (SS column output by VarScan) will NOT be used as a filter for this analysis since it is dependent on the somatic p-value, which we have optimized (and exceeds the standard signficance cutoff).

## ###############################################################################################
## Load and process datasets for paired T-N analysis (VarScan paired, MuTect, Strelka)
## ###############################################################################################

```{r}
library(dplyr)    # filter
library(tidyr)    # separate

#**************
#*  VARSCAN   *
#**************
varscan.paired.snvs <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/varScan.ME_or_blood_normals/all.paired_samples.varScan.snps.filtered.txt", sep = "\t", header = TRUE, fill = TRUE)
varscan.paired.indels <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/varScan.ME_or_blood_normals/all.paired_samples.varScan.indels.filtered.txt", sep = "\t", header = TRUE, fill = TRUE)

# combine snvs and indel - add mutation type as separate "type" column
varscan.paired.snvs$type = "snv"
varscan.paired.indels$type = "indel"
varscan.paired.calls <- rbind(varscan.paired.snvs, varscan.paired.indels)

rm(varscan.paired.snvs)
rm(varscan.paired.indels)

# rename columns
colnames(varscan.paired.calls) <- c("sample", "chr", "pos", "external_id", "ref", "alt", "qual", "filter",
                            "depth", "somatic_status", "ssc", "gpv", "somatic_p_value", "cda",
                            "KG_validated", "om", "pm", "gmaf", "gt_normal", "gt_tumor", "gq_normal", "gq_tumor",
                            "depth_normal", "depth_tumor", "ref_reads_normal", "ref_reads_tumor",
                            "var_reads_normal", "var_reads_tumor", "allele_freq_normal", "allele_freq_tumor",
                            "depth4_normal", "depth4_tumor", "effect", "impact", "fun_class", "HGVS_DNA",
                            "HGVS_protein", "gene", "biotype", "coding", "transcript_id", "exon_rank", "type")

# create a tumor_id column based on the sample id (format is <tumor_id>_<normal_id>) - e.g. GE0556B_GE0556-N
varscan.paired.calls$tumor_id <- gsub("(.*)\\_GZ(.*)","\\1", varscan.paired.calls$sample)
# similarly create a normal_id column
varscan.paired.calls$normal_id <- paste("GZ", gsub("(.*)\\_GZ(.*)","\\2", varscan.paired.calls$sample), sep="")

# remove ".markDups" from the ids
varscan.paired.calls$tumor_id <- gsub("(.*)\\.markDups(.*)","\\1", varscan.paired.calls$tumor_id)
varscan.paired.calls$normal_id <- gsub("(.*)\\.markDups(.*)","\\1", varscan.paired.calls$normal_id)
varscan.paired.calls$normal_id <- gsub("(.*)\\.picardMarkDups(.*)","\\1", varscan.paired.calls$normal_id)

# remove "%" from the allele frequency
varscan.paired.calls$allele_freq_normal <- gsub("(.*)\\%(.*)","\\1", varscan.paired.calls$allele_freq_normal)
varscan.paired.calls$allele_freq_tumor <- gsub("(.*)\\%(.*)","\\1", varscan.paired.calls$allele_freq_tumor)

########################
# Application of filters
########################

# define thresholds
MAX_GMAF = 0.01           # default
MAX_SOMATIC_PVAL = 0.175  # OPTIMIZED
MIN_VAR_READS_TUMOR = 3   # default
MIN_VAR_AF_TUMOR = 5      # default; as a percentage
MAX_VAR_AF_NORMAL = 15    # OPTIMIZED; as a percentage

# apply thresholds
varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(gmaf) < MAX_GMAF | is.na(gmaf))
varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(somatic_p_value) < MAX_SOMATIC_PVAL)
varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(var_reads_tumor) >= MIN_VAR_READS_TUMOR)
varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(allele_freq_tumor) >= MIN_VAR_AF_TUMOR)
varscan.paired.calls <- filter(varscan.paired.calls, as.numeric(allele_freq_normal) < MAX_VAR_AF_NORMAL)

# remove unneeded columns and rearrange as necessary
# last few columns are for initial filtering and reference and can be removed later
keeps <- c("tumor_id", "normal_id", "type", "chr", "pos", "gene", "ref", "alt", "HGVS_DNA", "HGVS_protein",
           "allele_freq_normal", "allele_freq_tumor", "depth4_normal", "depth4_tumor", 
           "var_reads_normal", "var_reads_tumor", "effect", "impact", "fun_class", "transcript_id",
           "external_id", "somatic_p_value", "somatic_status", "gt_normal", "gt_tumor", "gmaf")
varscan.paired.calls <- varscan.paired.calls[keeps]

# filter HGVS calls for obvious effects
varscan.paired.calls <- unique(rbind(
                varscan.paired.calls[grep("FRAME_SHIFT", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_ACCEPTOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("SPLICE_SITE_DONOR", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_CHANGE_PLUS_CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_DELETION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("CODON_INSERTION", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_CODING", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("NON_SYNONYMOUS_START", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("START_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_GAINED", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("STOP_LOST", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("UTR_3_PRIME", varscan.paired.calls$effect), ],
                varscan.paired.calls[grep("UTR_5_PRIME", varscan.paired.calls$effect), ]
                ))

# other effects to include:
# SPLICE_SITE_REGION ? [IMPACT: LOW] - desc: A sequence variant in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8 bases of the intron.
# EXON ? [IMPACT: MODIFIER] - desc: The variant hits an exon (from a non-coding transcript) or a retained intron.
# SYNONYMOUS_STOP? [IMPACT: LOW] - desc: Variant causes stop codon to be mutated into another stop codon. e.g.: taA/taG, */*

varscan.paired.calls$combine = as.character(paste(
  varscan.paired.calls$tumor_id, 
  varscan.paired.calls$chr, 
  varscan.paired.calls$pos, 
  varscan.paired.calls$ref,
  varscan.paired.calls$alt,
  sep = "."
  ))

# sort the calls by case, chr, then position
varscan.paired.calls <- arrange(varscan.paired.calls, tumor_id, chr, pos)

# print out the results (these will be VarScan not intersected with any other methods)
write.table(varscan.paired.calls, "/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/varScan.ME_or_blood_normals/varScan.all.paired-snps_and_indels.txt", sep = "\t", quote = FALSE, row.names = FALSE)

#**************
#*   MUTECT   *
#**************

# NB: the raw output file first needs to be cleaned up (includes extra headers from every tumor file), and will also include file names from performing an egrep on multiple files.  This can be done quickly in the command line environment.
mutect.snvs <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/mutect/mutect.all.snps.pass.txt", sep = "\t", header = TRUE)

mutect.snvs$combine = as.character(paste(
  mutect.snvs$tumor_name,
  mutect.snvs$contig,
  mutect.snvs$position,
  mutect.snvs$ref_allele,
  mutect.snvs$alt_allele,
  sep = "."
))

mutect.snvs$type = "snv"

#**************
#*  STRELKA   *
#**************
strelka.snvs <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/strelka/passed.snvs.txt", sep = "\t", header = FALSE)
strelka.indels <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/strelka/passed.indels.txt", sep = "\t", header = FALSE)

# combine snvs and indels - add mutation type as separate "type" column
strelka.indels$type = "indel"
strelka.snvs$type = "snv"
strelka.calls <- rbind(strelka.indels, strelka.snvs)

rm(strelka.indels)
rm(strelka.snvs)

# extract only columns of interest
strelka.calls <- strelka.calls[,c("V1", "V2", "V3", "V5", "V6", "type")]

# assign column names to strelka datasets
colnames(strelka.calls) <- c("sample", "chr", "pos", "ref", "alt", "type")

# remove ".markDups" from the tumor ids
strelka.calls$sample <- gsub("(.*)\\.markDups(.*)","\\1", strelka.calls$sample)

strelka.calls$combine = as.character(paste(
  strelka.calls$sample,
  strelka.calls$chr,
  strelka.calls$pos,
  strelka.calls$ref,
  strelka.calls$alt,
  sep = "."
))

# Find the intersection of varscan (snvs), mutect, and strelka (snvs).
combine.varscan.intersect.strelka <- intersect(unique(varscan.paired.calls$combine), unique(strelka.calls$combine))
combine.intersect.all <- intersect(unique(combine.varscan.intersect.strelka), unique(mutect.snvs$combine))
snvs.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.intersect.all)
snvs.intersect.all$evidence <- "mutect, strelka, varscan"
#
# add indels
# Take the intersect of indels from varscan and strelka indel calls.
varscan.indels <- subset(varscan.paired.calls, varscan.paired.calls$type == 'indel')
strelka.indels <- subset(strelka.calls, strelka.calls$type == 'indel')
combine.varscan.intersect.strelka <- intersect(unique(varscan.indels$combine), unique(strelka.indels$combine))
indels.intersect.all <- subset(varscan.paired.calls, varscan.paired.calls$combine %in% combine.varscan.intersect.strelka)
indels.intersect.all$evidence <- "strelka, varscan"

# now combine snvs and indels
calls.all <- rbind(snvs.intersect.all, indels.intersect.all)

rm(varscan.paired.calls)
rm(strelka.calls)
rm(mutect.snvs)
rm(combine.varscan.intersect.strelka)
#rm(combine.intersect.all)
#rm(snvs.intersect.all)
#rm(varscan.indels)
#rm(strelka.indels)
#rm(indels.intersect.all)

# remove NA's and blanks for functional class
calls.all$fun_class[calls.all$fun_class==""] <- "NONE"
calls.all$fun_class[is.na(calls.all$fun_class)] <- "NONE"

# remove the "combine" and "tumor_id" column, and rearrange columns in the process
keeps <- c("tumor_id", "normal_id", "type", "chr", "pos", "gene", "ref", "alt", "HGVS_DNA", "HGVS_protein",
           "allele_freq_normal", "allele_freq_tumor", "depth4_normal", "depth4_tumor", 
           "var_reads_normal", "var_reads_tumor", "effect", "impact", "fun_class", "transcript_id",
           "external_id", "somatic_p_value", "somatic_status", "gt_normal", "gt_tumor", "gmaf", "evidence")

calls.all <- calls.all[keeps]

# sort the calls by case, chr, then position
calls.all <- unique(arrange(calls.all, tumor_id, chr, pos))

# write the results to a text file
write.table(calls.all, "/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/snvs.all.intersect-ME_or_blood_normal.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```

## ###############################################################################################
## Visualization
## ###############################################################################################

```{r}
library(plyr)
library(ggplot2)

calls.ME_or_blood_normals.OPTIMIZED <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/snvs.all.intersect-ME_or_blood_normal-OPTIMIZED_thresholds.txt", sep = "\t", header = TRUE, fill = TRUE)
calls.ME_or_blood_normals.DEFAULT <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/optimized_thresholds-full_exome_cohort/snvs.all.intersect-ME_or_blood_normal-DEFAULT_thresholds.txt", sep = "\t", header = TRUE, fill = TRUE)

calls.ME_or_blood_normals.OPTIMIZED$class <- revalue(calls.ME_or_blood_normals.OPTIMIZED$effect, c("NON_SYNONYMOUS_CODING"="Missense", 
                           "NON_SYNONYMOUS_START"="Missense",
                           "START_LOST"="Start Lost",
                           "STOP_LOST"="Missense",
                           "STOP_GAINED"="Nonsense",
                           "CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "CODON_CHANGE_PLUS_CODON_INSERTION"="In-frame Indel",
                           "CODON_DELETION"="In-frame Indel",
                           "CODON_INSERTION"="In-frame Indel",
                           "FRAME_SHIFT"="Frameshift Indel",
                           "FRAME_SHIFT+SPLICE_SITE_ACCEPTOR"="Frameshift Indel",
                           "FRAME_SHIFT+SPLICE_SITE_DONOR"="Frameshift Indel",
                           "FRAME_SHIFT+STOP_GAINED"="Frameshift Indel",
                           "SPLICE_SITE_ACCEPTOR+CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "SPLICE_SITE_DONOR+CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "STOP_LOST+CODON_DELETION"="In-frame Indel",
                           "SPLICE_SITE_ACCEPTOR"="Splice site",
                           "SPLICE_SITE_DONOR"="Splice site",
                           "UTR_3_PRIME"="3' UTR",
                           "UTR_5_PRIME"="5' UTR"
                           ))

calls.ME_or_blood_normals.DEFAULT$class <- revalue(calls.ME_or_blood_normals.DEFAULT$effect, c("NON_SYNONYMOUS_CODING"="Missense", 
                           "NON_SYNONYMOUS_START"="Missense",
                           "START_LOST"="Start Lost",
                           "STOP_LOST"="Missense",
                           "STOP_GAINED"="Nonsense",
                           "CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "CODON_CHANGE_PLUS_CODON_INSERTION"="In-frame Indel",
                           "CODON_DELETION"="In-frame Indel",
                           "CODON_INSERTION"="In-frame Indel",
                           "FRAME_SHIFT"="Frameshift Indel",
                           "FRAME_SHIFT+SPLICE_SITE_ACCEPTOR"="Frameshift Indel",
                           "FRAME_SHIFT+SPLICE_SITE_DONOR"="Frameshift Indel",
                           "FRAME_SHIFT+STOP_GAINED"="Frameshift Indel",
                           "SPLICE_SITE_ACCEPTOR+CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "SPLICE_SITE_DONOR+CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "STOP_LOST+CODON_DELETION"="In-frame Indel",
                           "SPLICE_SITE_ACCEPTOR"="Splice site",
                           "SPLICE_SITE_DONOR"="Splice site",
                           "UTR_3_PRIME"="3' UTR",
                           "UTR_5_PRIME"="5' UTR"
                           ))


data.df <- as.data.frame(table(calls.ME_or_blood_normals.OPTIMIZED$tumor_id, calls.ME_or_blood_normals.OPTIMIZED$class))
colnames(data.df) <- c("sample", "variant_class", "count")

p <- ggplot(data = data.df, aes(x = sample, y = count, fill = variant_class)) + 
  geom_bar(stat = "identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations")

#geom_text(aes(label=count), vjust=-0.25) +

# same plot, but without UTR mutations
UTR_mutations <- c("UTR_3_PRIME", "UTR_5_PRIME")
calls.subset <- subset(calls.ME_or_blood_normals.OPTIMIZED, !(calls.ME_or_blood_normals.OPTIMIZED$effect %in% UTR_mutations))
calls.subset$effect <- factor(calls.subset$effect)
calls.subset$class <- factor(calls.subset$class)
#calls.subset$effect <- factor(calls.subset$effect)
data.df <- as.data.frame(table(calls.subset$tumor_id, calls.subset$class))
colnames(data.df) <- c("sample", "variant_class", "count")

p <- ggplot(data = data.df, aes(x = sample, y = count, fill = variant_class)) + 
  geom_bar(stat = "identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations")

# barplot comparing total mutations per sample between default and optimized datasets
num_calls.optimized <- as.data.frame(table(calls.ME_or_blood_normals.OPTIMIZED$tumor_id))
num_calls.optimized$thresholds <- "Optimized"
num_calls.default <- as.data.frame(table(calls.ME_or_blood_normals.DEFAULT$tumor_id))
num_calls.default$thresholds <- "Default"

num_calls.combined <- rbind(num_calls.default, num_calls.optimized)
colnames(num_calls.combined) <- c("sample", "num_mutations", "thresholds")

p <- ggplot(data = num_calls.combined, aes(x = sample, y = num_mutations, fill = thresholds)) + 
  geom_bar(stat = "identity", width = 0.6, position = position_dodge()) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_discrete(name = "Thresholds") +
  xlab("") + ylab("Number of mutations")


```

## Allele frequency

```{r}
library(gridExtra)
library(gtable)
library(ggplot2)

# for each sample, create AF density plot with vertical at peak value
AF.plots <- vector('list', length(unique(calls.all$tumor_id)))
for (i in seq_along(unique(calls.all$tumor_id))) {
  temp <- subset(calls.all, calls.all$tumor_id == unique(calls.all$tumor_id)[i])
  
  num_mutations <- nrow(temp)
  
  # identify peak value for current sample
  max_y_density <- which.max(density(temp$allele_freq_tumor)$y)
  max_x_intercept <- density(temp$allele_freq_tumor)$x[max_y_density]
  max_y_intercept <- density(temp$allele_freq_tumor)$y[max_y_density]
  
  print(max_y_intercept)
  
  AF.plots[[i]] <- ggplot(temp, aes(as.numeric(allele_freq_tumor))) + 
    geom_density() +
    geom_vline(xintercept = max_x_intercept, col="red", lty=2) +
    xlab("Allele frequency (%)") + ylab("Density") +
    annotate("text", 
             y = max_y_intercept+0.2*max_y_intercept,
             x = max(temp$allele_freq_tumor),
             label = paste("Peak AF: ", round(max_x_intercept, digits = 2), "%", sep = ""),
             hjust = 1) + 
    theme(axis.title.x = element_blank(), axis.title.y = element_blank()) +
    ylim(0, max_y_intercept+0.25*max_y_intercept) +
    ggtitle(label = paste(temp$tumor_id, " (n = ", num_mutations, ")", sep = ""))
}
#do.call("grid.arrange", c(AF.plots, ncol=1))
grid.arrange(grobs = AF.plots, ncol=3, bottom = "Allele frequency (%)", left = "Density")

#y = max(density(temp$allele_freq_tumor)$y),

# plot all samples
p <- ggplot(calls.all, aes(as.numeric(allele_freq_tumor), colour = tumor_id)) + 
  geom_density() + 
  xlab ("Allele frequency (%)") + 
  ylab("Density")

# faceted plot (one distribution per sample)
p + facet_grid(tumor_id ~ ., scales = "free")

# or organize plots into 2 columns
p + facet_wrap(~ tumor_id, ncol=2)

```
