---
title: "Analysis of mutations in GZL exomes"
author: "Stacy Hung"
output:
  pdf_document: default
  html_notebook: default
---

## ###############################################################################################
## Based on SNPs and indels called in all normals (n=37), how many normals does a "target" variant 
## appear in ?  More than 2 or 3 would suggest that the variant is actually a germline variant.
## ###############################################################################################

```{r}
library(plyr) # for revalue function
library(dplyr)    # filter
library(hashmap) # for hashmap function

###############################
# Read in data related to tumor
###############################

# read in variants for full exome cohort
calls.part1 <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/snv_indel_datasets/snvs_indels.30_final.default_and_optimized.txt", sep = "\t", header = TRUE, fill = TRUE)
calls.part2 <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/snv_indel_datasets/t_v_me.all/snvs_indels.t_v_me.GZ197.OPTIMIZED_thresholds.txt", sep = "\t", header = TRUE, fill = TRUE)
calls.all <- rbind(calls.part1, calls.part2)
rm(calls.part1)
rm(calls.part2)

# read in target genes
target.genes <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/normal_SNP_analysis/target_genes.FINAL_216.txt", header = FALSE)
colnames(target.genes) <- c("gene")

# read in reference table with tumor-normal pairs and corresponding peak VAFs
ref.table <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/WES_data/all.tumors_normals-peak_VAF_and_est_tumor_content.txt", sep = "\t", header = TRUE)

# filter list of variants present only in target genes
target.calls <- filter(calls.all, calls.all$gene %in% target.genes$gene)

###############################
# Read in data related to normal
###############################

# read in normal SNPs (only need sample, chrom, pos, ref, and alt positions)
# NB: this variant list was obtained by running VarScan on all 37 normals with PVAL=1 and MIN_VARIANT_READS=3 as the only thresholds
snps.normals <- unique(read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/normal_SNP_analysis/varScan/all.single_samples.varScan.snps.snpEff_canonical.filtered.txt", sep = "\t", header = TRUE, fill = TRUE))
colnames(snps.normals) <- c("normal_id", "chr", "pos", "ref", "alt", "allele_freq", "gene")
snps.normals$allele_freq <- gsub("(.*)\\%(.*)","\\1", snps.normals$allele_freq)

# read in normal indels 
indels.normals <- unique(read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/normal_SNP_analysis/varScan/all.single_samples.varScan.indels.snpEff_canonical.indels_annotated.cut.txt", sep = "\t", header = TRUE, fill = TRUE))
colnames(indels.normals) <- c("normal_id", "chr", "pos", "ref", "alt", "allele_freq", "gene")
indels.normals$allele_freq <- gsub("(.*)\\%(.*)","\\1", indels.normals$allele_freq)

# combine snps and indels
calls.normals <- rbind(snps.normals, indels.normals)
rm(snps.normals)
rm(indels.normals)

# filter for SNPs that occur in target genes
calls.normals <- filter(calls.normals, calls.normals$gene %in% target.genes$gene)

#write.table(calls.normals, "/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/normal_SNP_analysis/to_validate_all/snps_and_indels.target_variants.txt", sep = "\t", quote = FALSE, row.names = FALSE)


#######################################
# Create indices, etc. for fast mapping
#######################################

target.calls$key <- as.character(paste(
  target.calls$chr,
  target.calls$pos,
  target.calls$ref,
  target.calls$alt,
  sep = "."
))

# create keys to cross-reference: "chr", "pos", "ref", "alt"
calls.normals$key <- as.character(paste(
  calls.normals$chr,
  calls.normals$pos,
  calls.normals$ref,
  calls.normals$alt,
  sep = "."
))

# extract germline variants that had a hit in the target list:
calls.normals <- filter(calls.normals, calls.normals$key %in% target.calls$key)

write.table(calls.normals, "/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/normal_SNP_analysis/final_pipeline/calls.normals-target.genes.txt", sep = "\t", quote = FALSE, row.names = FALSE)

# create a table of snps
snps.counts <- as.data.frame(table(calls.normals$key))
colnames(snps.counts) <- c("snp", "num_normals")
hash.snp.counts <- hashmap(as.character(snps.counts$snp), as.numeric(snps.counts$num_normals))

###########################
# Apply filtering algorithm
###########################

target.calls$num_normals <- 0
target.calls$normal_snp_filter <- 0

# for each target variant; add data for 2 new columns: normal_SNP_filter, and # normals variant is present in

# now cross reference target calls with snps 
for (i in 1:nrow(target.calls)){
  # check if the target variant occurs in any of the normals
  num_normals <- hash.snp.counts[[target.calls[i, ]$key]]
  
  # store as a number
  if (is.na(num_normals)){
    num_normals <- 0
  }
  target.calls[i, ]$num_normals <- num_normals
  
  # if variant is in at least 1 normal that does not include the matching normal
  if (num_normals > 0){
    
    # get list of normal_ids
    normal.ids <- subset(calls.normals, calls.normals$key == target.calls[i, ]$key)$normal_id
    normal.ids <- gsub("(.*)\\.markDups(.*)","\\1", normal.ids)
    # check for matching normal
    if (target.calls[i, ]$normal_id %in% normal.ids) {
      num_normals <- num_normals - 1
      target.calls[i, ]$num_normals <- num_normals
      # now apply filters with updated normal hits
      if (num_normals >= 2){
        target.calls[i, ]$normal_snp_filter <- "EXCLUDE; hit to matching normal not included in analysis"
      } else {
        # else variant must be present in just 1 normal (not matching)
        # isolate non-matching normal
        normal.ids <- filter(normal.ids, !(normal.ids %in% target.calls[i, ]$normal_id))
        # get VAF of normal SNP (exclude matching normal)
        normals <- subset(calls.normals, calls.normals$key == target.calls[i, ]$key)
        normal.non_matching <- filter(normals, !(normals$normal_id %in% target.calls[i, ]$normal_id))
        normal.VAF <- as.numeric(normal.non_matching$allele_freq)
        # get peak VAF for tumor matching normal with SNP
        tumor.peak_VAF <- as.numeric(subset(ref.table, ref.table$normal_id == as.character(normal.non_matching$normal_id))$peak_VAF)
        # if (VAF normal > peak VAF)
        if (normal.VAF > tumor.peak_VAF){
          target.calls[i, ]$normal_snp_filter <- paste("EXCLUDE; hit to matching normal not included in analysis; VAF ", as.character(normal.non_matching$normal_id), " (", normal.VAF, ") > Peak VAF in matching tumor (", tumor.peak_VAF, ")",  sep = "")
        } else {
          # else, normal_SNP_filter = "KEEP; 1 normal"          
          target.calls[i, ]$normal_snp_filter <- "KEEP"
        }
      }
    } else {
      # variant is present in 1 or more normals, none of which are matching
      if (num_normals >= 2){
        target.calls[i, ]$normal_snp_filter <- "EXCLUDE"
      } else {
        # else variant must be present in just 1 normal (not matching)
        # get VAF of normal SNP
        normal.VAF <- as.numeric(subset(calls.normals, calls.normals$key == target.calls[i, ]$key)$allele_freq)
        # get peak VAF for tumor matching normal with SNP
        tumor.peak_VAF <- as.numeric(subset(ref.table, ref.table$normal_id == as.character(normal.ids[[1]]))$peak_VAF)
        # if (VAF normal > peak VAF)
        if (normal.VAF > tumor.peak_VAF){
          target.calls[i, ]$normal_snp_filter <- paste("EXCLUDE; VAF ", as.character(normal.ids[[1]]), "(", normal.VAF, ") > Peak VAF in matching tumor (", tumor.peak_VAF, ")",  sep = "")
        } else {
          # else, normal_SNP_filter = "KEEP; 1 normal"          
          target.calls[i, ]$normal_snp_filter <- "KEEP"
        }
      }
    }
  } else {
    target.calls[i, ]$normal_snp_filter <- "KEEP"
  }
  

}

# write out final results
write.table(target.calls, "/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/snv_indel_datasets/snvs_indels.31_final.default_and_optimized.num_normals.all_targets.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```