---
title: "SNV Heatmap"
author: "Stacy Hung"
date: "March 26, 2019"
output: html_document
---

This script generates an oncoplot that orders mutations by pathology and for genes that have been selected accordingly:
  * Gene must be present in the target gene panel (n=217)
    * Gene is mutated in 10%+ of group 0 OR
    * Gene is mutated in 10%+ of group 1 and 2 combined OR
    * Gene is mutated in group 3
    
Mutations have been filtered as follows:
  - somatic effect, excluding UTR mutations
  - filters used for SNV/indel analysis pipeline (default filters for t vs. blood and optimized filters for t vs. ME)
  
NB: Previously, genes were filtered based on the requirement of being mutated in 2+ cases; in this case, we do not want to apply this filter since we want a bigger picture across pathologies.

## Load packages

```{r}
# NB: if you need to install this package, install it using devtools from the github source!!!
# The bioconductor version of GenVisR is buggy (not compatible with required libraries)
# Installing GenVisR may also require the installation of the VariantAnnotation pacakge (from bioconductor)
source("https://bioconductor.org/biocLite.R")
biocLite("VariantAnnotation")
library("VariantAnnotation")

install.packages("devtools")
library(devtools)
install_github("griffithlab/GenVisR")
library("GenVisR")

source("https://bioconductor.org/biocLite.R")
biocLite("maftools")
library(maftools)

library(gridExtra)
library(dplyr)
library(plyr)
```

## Transform GISTIC data into MAF format so that it can be used for oncoplot

```{r}
library(tidyr)

gistic.heatmap <- read.table("/Volumes/shung/projects/gzl_exomes/gistic/output/0.3_amp_0.3_del_0.98_focal_cutoff_2.5_cap_FINAL/for_R_oncoplot/selected_peaks_for_R_oncoplot.txt", sep = "\t", header = TRUE)

# wide to long
gistic.heatmap.long <- gather(gistic.heatmap, case, copy_number_change, GZ.BCC.020:GZ.229, factor_key = TRUE)

# convert to maf format (add appropriate columns as needed)
colnames(gistic.heatmap.long) <- c("Hugo_Symbol", "Tumor_Sample_Barcode", "Variant_Classification")

write.table(gistic.heatmap.long, "/Volumes/shung/projects/gzl_exomes/gistic/output/0.3_amp_0.3_del_0.98_focal_cutoff_2.5_cap_FINAL/for_R_oncoplot/gistic.heatmap.long.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```

## GISTIC oncoplot

```{r}
library(maftools)

#mut.df <- read.table("/Volumes/shung/projects/gzl_exomes/gistic/output/0.3_amp_0.3_del_0.98_focal_cutoff_2.5_cap_FINAL/for_R_oncoplot/selected_gistic_peaks_v2.maf", sep = "\t", header = TRUE)
mut.df <- read.table("/Volumes/shung/projects/gzl_exomes/gistic/output/0.3_amp_0.3_del_0.98_focal_cutoff_2.5_cap_FINAL/for_R_oncoplot/test4.maf", sep = "\t", header = TRUE)

# require fields: Hugo_Symbol, Tumor_Sample_Barcode, Variant_Classification, Chromosome, Start_Position, End_Position,  Tumor_Seq_Allele2, Reference_Allele, Variant_Type (only the first 3 fields need to have values)
mutations.maf <- read.maf("/Volumes/shung/projects/gzl_exomes/gistic/output/0.3_amp_0.3_del_0.98_focal_cutoff_2.5_cap_FINAL/for_R_oncoplot/selected_gistic_peaks_v3.maf", 
                          vc_nonSyn = c("High-level amplification", "Low-level amplification", 
                                        "High-level deletion", "Low-level deletion"))

# for oncoplot - want to display patients in same order as oncoplot-by_path_group.pdf
patients.ordered <- read.table("/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/patients.ordered.CURRENT.txt")
patients.ordered <- as.vector(patients.ordered$x)

# for feature ordering (in this case, the features are gistic peaks with GOI), first order by CNA type (amplifications, then losses) and then by decreasing frequency across the cohort.
features.ordered <- c("JAK2, CD274, PDCD1LG2 (9p24.1)", "USP18 (22q11.21)", "GTF2H2, NAIP (5q13.2)", "REL, BCL11A (2p16.1)", "SLX1B (16p11.2)", "PRMT6 (1p21.1)", "IRF5, RELN, LRRN3 (7q35)", "ATM (11q14.3)", "NCOR2 (12q24.33)", "CDKN2A, CDKN2B (9p21.3)")
#features.ordered <- c("9p24.1", "22q11.21", "5q13.2", "2p16.1", "16p11.2", "1p21.1", "7q35", "11q14.3", "12q24.33", "9p21.3")
#features.ordered <- c("JAK2, CD274, PDCD1LG2", "USP18", "GTF2H2, NAIP", "REL, BCL11A", "SLX1B", "PRMT6", "IRF5, RELN, LRRN3", "ATM", "NCOR2", "CDKN2A, CDKN2B")

# change colours for variant classifications
# light read --> R:243, G:191, B:181 or hex: #F3BFBF
#vc_cols <- c("firebrick", "firebrick1", "grey", "slategrey3", "steelblue4")
#names(vc_cols) <- c("Amplification", "Gain", "No change", "Heterozygous loss", "Homozygous loss")
vc_cols <- c("firebrick", "#F3BFBF", "slategray3", "steelblue4")
names(vc_cols) <- c("High-level amplification", "Low-level amplification", "Low-level deletion", "High-level deletion")

oncoplot(maf = mutations.maf, drawRowBar = TRUE,
         sampleOrder = patients.ordered,
         genes = features.ordered, keepGeneOrder = TRUE,
         colors = vc_cols,
         legendFontSize = 1.5, SampleNamefontSize = 0.5, fontSize = 0.5,
         showTumorSampleBarcodes = TRUE, 
         titleFontSize = 0.01)

# for future debugging: need to figure out why frequency bars on right-hand side are not showing up

```

## Get list of genes to use for oncoplot

```{r}
library(dplyr)
library(plyr)

# get full list of mutations that will be potentially used for oncoplot (excludes GZ2229 only)
mutations.df <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/snv_indel_datasets/maf/gzl_exomes_incl_GZ222-snvs_indels.maf", sep = "\t", header = TRUE)

# load target panel genes
genes.target <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/WES_data/target_genes.FINAL_217.txt", header = FALSE)
colnames(genes.target) <- c("gene")
genes.target <- as.vector(genes.target$gene)

# load metadata
clinical.data <- read.table("/Volumes/shung/projects/gzl_exomes/data/GZ_WES_cases-clinical_data.mapped.txt", sep = "\t", header = TRUE)

# filter mutation dataset for target genes --> there are 151 genes mutated that are also present on the target panel
mutations.df <- filter(mutations.df, mutations.df$Hugo_Symbol %in% genes.target)
mutations.df$Hugo_Symbol <- factor(mutations.df$Hugo_Symbol)

# map on formal patient id mappings
patient_id.to.new_patient_id <- as.vector(clinical.data$Tumor_Sample_Barcode)
names(patient_id.to.new_patient_id) <- as.vector(clinical.data$tumor_id2)
mutations.df$Tumor_Sample_Barcode <- revalue(mutations.df$Tumor_Sample_Barcode, patient_id.to.new_patient_id)

# map on pathology group (group 0, 1+2, 3)
patient.to.path_group <- as.vector(clinical.data$Path_group)
names(patient.to.path_group) <- as.vector(clinical.data$Tumor_Sample_Barcode)
mutations.df$Path_group <- revalue(mutations.df$Tumor_Sample_Barcode, patient.to.path_group)

# calculate NUMBER OF CASES mutated in each GENE
mutated_cases.by_path_and_gene <- unique(mutations.df[c("Hugo_Symbol", "Tumor_Sample_Barcode", "Path_group")])
mutated.table <- as.data.frame(table(mutated_cases.by_path_and_gene$Path_group, mutated_cases.by_path_and_gene$Hugo_Symbol))
colnames(mutated.table) <- c("Path_group", "gene", "num_mutated")

mutated.group_0 <- unique(filter(mutated_cases.by_path_and_gene, mutated_cases.by_path_and_gene$Path_group == "0 = cHL"))
mutated.group_1_2 <- unique(filter(mutated_cases.by_path_and_gene, (mutated_cases.by_path_and_gene$Path_group == "1 = cHL-interm" | 
                                               mutated_cases.by_path_and_gene$Path_group == "2 = LBCL-interm")))
mutated.group_3 <- unique(filter(mutated_cases.by_path_and_gene, mutated_cases.by_path_and_gene$Path_group == "3 = LBCL"))

# get NUMBER OF CASES mutated per GROUP (to calculate normalized proportions)
NUM_GROUP_0 <- length(unique(mutated.group_0$Tumor_Sample_Barcode))
NUM_GROUP_1_2 <- length(unique(mutated.group_1_2$Tumor_Sample_Barcode))
NUM_GROUP_3 <- length(unique(mutated.group_3$Tumor_Sample_Barcode))

# extract GENES mutated in each GROUP 
num_mutated.by_gene.group_0 <- subset(mutated.table, mutated.table$Path_group == "0 = cHL" & mutated.table$num_mutated > 0)
num_mutated.by_gene.group_1_and_2 <- subset(mutated.table, (mutated.table$Path_group == "1 = cHL-interm" | 
                                       mutated.table$Path_group == "2 = LBCL-interm") & mutated.table$num_mutated > 0)
# combine counts for group 1 and 2, since they are currently separate
num_mutated.group_1_2.combined_counts <- as.data.frame(table(mutated.group_1_2$Hugo_Symbol))
colnames(num_mutated.group_1_2.combined_counts) <- c("gene", "num_mutated")
num_mutated.by_gene.group_3 <- subset(mutated.table, mutated.table$Path_group == "3 = LBCL" & mutated.table$num_mutated > 0)

# calculate proportion of mutated cases
num_mutated.by_gene.group_0$proportion <- num_mutated.by_gene.group_0$num_mutated / NUM_GROUP_0*100
num_mutated.group_1_2.combined_counts$proportion <- num_mutated.group_1_2.combined_counts$num_mutated / NUM_GROUP_1_2*100

PERCENT_MUTATED = 10

# identify genes that fall are present in 10%+ of group 0, genes present in 10%+ of group 1+2, and all genes mutated in group 3
num_mutated.pathgp.0_cutoff <- subset(num_mutated.by_gene.group_0, num_mutated.by_gene.group_0$proportion >= PERCENT_MUTATED)
num_mutated.pathgp.1_2_cutoff <- subset(num_mutated.group_1_2.combined_counts, 
                                        num_mutated.group_1_2.combined_counts$proportion >= PERCENT_MUTATED)
pathgp_genes.10_plus <- unique(c(as.vector(num_mutated.pathgp.0_cutoff$gene),as.vector(num_mutated.pathgp.1_2_cutoff$gene),
                                 as.vector(num_mutated.by_gene.group_3$gene)))

write.table(pathgp_genes.10_plus, "/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/pathgp_genes.10_plus-v_incl_GZ222.txt", quote = FALSE, row.names = FALSE)

# create mutation matrix for filtered gene list
#mutations.df <- read.table("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/snv_indel_datasets/maf/gzl_exomes-snvs_indels.mutated_target_panel.maf", sep = "\t", header = TRUE, fill = TRUE)
mutations.df <- filter(mutations.df, mutations.df$Hugo_Symbol %in% pathgp_genes.10_plus)
write.table(mutations.df, "/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/snv_indel_datasets/maf/gzl_exomes_incl_GZ222-snvs_indels.filtered_path_group.maf", sep = "\t", quote = FALSE, row.names = FALSE)

mutations.df$Hugo_Symbol <- factor(mutations.df$Hugo_Symbol)
mutations.df <- unique(mutations.df[c("Hugo_Symbol", "Tumor_Sample_Barcode")])
mutation.matrix <- table(mutations.df$Tumor_Sample_Barcode, mutations.df$Hugo_Symbol)
write.table(mutation.matrix, "/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/mutation_matrix_incl_GZ222-path_groups_10+_freq_mutated.txt", sep = "\t", quote = FALSE)
```

## Sort matrix

```{r}
# for each gene, calculate number of mutated cases; and number of pathology group-specific cases
#mut.data <- read.table("/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/mutation_matrix-path_groups_10+_freq_mutated.with_path_group_and_sums.txt", sep = "\t", header = TRUE)
mut.data <- read.table("/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/mutation_matrix_incl_GZ222-path_groups_10+_freq_mutated.with_path_group_and_sums.txt", sep = "\t", header = TRUE)

# sort by each gene in their current order
#mut.data.sorted <- mut.data[with(mut.data, order(Path_group, -SOCS1, -B2M, -GNA13, -LRRN3, -CSMD3, -CSF2RB, -ITPKB, -NCOR1, -ZNF217, -IRF2BP2, -ARID1A, -EGR1, -IRF4, - LMF2, -PRKDC, -RELN, - SPEN, -ABCA13, -ABCC6, -C2CD3, -CD58, - CISH, - KCNQ3, -MFHAS1, -MLL2, -NFKBIE, -RC3H2, -SIAH2 , -STAT6, -TNFAIP3, -WEE1, -DHX58, -GRIA4, -ABCD2, -ADAP1, -AMOT, - BCL2, - BRWD3, -GRIN3A, -HIST1H1B, -MAPK6, -MCL1, -MSH6, -SETD1B, -SHISA6, -SPSB4, -USP5, -XCR1, -DIAPH3, -ARID5B, -FLT4, -NLRP4, -TAB1, -HIST1H1E, -NFKBIB, -POU2AF1, -ATM, -SPOP)), ]

mut.data.sorted <- mut.data[with(mut.data, order(Path_group, -SOCS1, -B2M, -GNA13, -LRRN3, -CSMD3, -CSF2RB, -ITPKB, -NCOR1, -NFKBIE, -ZNF217, -IRF2BP2, -ARID1A, -EGR1, -IRF4, -LMF2, -PRKDC, -RELN, -SPEN, -ABCA13, -ABCC6, -C2CD3, -CD58, -CISH, -KCNQ3, -MFHAS1, -MLL2, -RC3H2, -SIAH2, -STAT6, -TNFAIP3, -WEE1, -DHX58, -GRIA4, -ABCD2, -ADAP1, -AMOT, -BCL2, -BRWD3, -GRIN3A, -HIST1H1B, -HIST1H1E, -MAPK6, -MCL1, -MSH6, -SETD1B, -SHISA6, -SPSB4, -USP5, -XCR1, -DIAPH3, -ARID5B, -ATM, -FLT4, -NLRP4, -TAB1, -NFKBIB, -POU2AF1, -SPOP)), ]

patients.ordered <- as.vector(mut.data.sorted$patient)
write.table(patients.ordered, "/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/patients.ordered.CURRENT.txt", sep = "\t", quote = FALSE)

mut.oncoplot <- t(mut.data.sorted)
genes.ordered <- as.vector(row.names(mut.oncoplot))
genes.ordered <- genes.ordered[ genes.ordered != "patient"]
genes.ordered <- genes.ordered[ genes.ordered != "Path_group"]
write.table(genes.ordered, "/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/genes.ordered.CURRENT.txt", sep = "\t", quote = FALSE)

write.table(mut.oncoplot, "/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/mutation_matrix_incl_GZ222-path_groups_10+_freq_mutated.oncoplot.txt", sep = "\t", quote = FALSE)

```

## Path group-specific oncoplots

```{r}
# to get latest version of maftools, install from github
library("devtools")
install_github(repo = "PoisonAlien/maftools")
library(maftools)

clinical.data <- read.table("/Volumes/shung/projects/gzl_exomes/data/GZ_WES_cases-clinical_data.with_updated_mediastinal_and_age.txt", sep = "\t", header = TRUE)
clinical.data$Path_group <- revalue(clinical.data$Path_group, 
                                          c("0 = cHL"="0_=_cHL",
                                            "1 = cHL-interm"="1_=_cHL-interm",
                                            "2 = LBCL-interm"="2_=_LBCL-interm",
                                            "3 = LBCL"="3_=_LBCL"))

#mutations.maf <- read.maf("/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/gzl_exomes-snvs_indels.filtered_path_group.with_REMOVE_gene.maf", clinicalData = clinical.data)
mutations.maf <- read.maf("/Volumes/shung/projects/gzl_exomes/snv_analysis-20181127/snv_indel_datasets/maf/gzl_exomes_incl_GZ222-snvs_indels.filtered_path_group.with_REMOVE_gene.maf", clinicalData = clinical.data)

genes.ordered <- read.table("/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/genes.ordered.CURRENT.txt")
genes.ordered <- as.vector(genes.ordered$x)
patients.ordered <- read.table("/Volumes/shung/projects/gzl_exomes/data/Path_group_visualization/patients.ordered.CURRENT.txt")
patients.ordered <- as.vector(patients.ordered$x)

# colours for annotation tracks
annot.colors <- list(Gender = c("Female" = "palevioletred1", "Male" = "royalblue1"),
                     EBER = c("Negative" = "black", "Positive" = "red"),
                     Mediastinal = c("Not_involved" = "grey", "Secondary" = "plum2", "Primary" = "purple"),
                     Age_yrs = c("<_45" = "bisque", "45_<_age_<_60"="navajowhite3", ">=_60" = "bisque4"),
                     Path_group = c("0_=_cHL" = "lightsalmon",
                                   "1_=_cHL-interm" = "khaki1",
                                   "2_=_LBCL-interm" = "lightgreen", 
                                   "3_=_LBCL" = "steelblue1")
               )

# show patients ordered by group, and mutations ordered by frequency
# include "REMOVE" gene since the removeNonMutated parameter doesn't work (i.e. doesn't show non-mutated case when FALSE)
genes.ordered <- c(genes.ordered, "REMOVE")

oncoplot(maf = mutations.maf, SampleNamefontSize = 0.6,
         removeNonMutated = FALSE,
         sampleOrder = patients.ordered,
         genes = genes.ordered, 
         keepGeneOrder = TRUE, fontSize = 0.6, titleFontSize = 0, legendFontSize = 1, annotationFontSize = 1,
         clinicalFeatures = c("Path_group", "EBER", "Mediastinal_involvement", "Gender", "Age_category"),
         annotationColor = annot.colors, 
         showTumorSampleBarcodes = TRUE,
         drawColBar = FALSE)

# version with large patient font
oncoplot(maf = mutations.maf, SampleNamefontSize = 0.75,
         removeNonMutated = FALSE,
         sampleOrder = patients.ordered, 
         genes = genes.ordered, 
         keepGeneOrder = TRUE, fontSize = 0.6, titleFontSize = 0, legendFontSize = 0, annotationFontSize = 0,
         showTumorSampleBarcodes = TRUE,
         drawColBar = FALSE)

```

